File: ios/RNBleError.swift
//
//  RNBleError.swift
//  RNDsbluetooth
//
//  Created by Jaysonj on 2019/2/18.
//  Copyright © 2019 Facebook. All rights reserved.
//

import Foundation

enum BleErrorCode : Int {
    case UnknownError = 0
    case BluetoothManagerDestroyed = 1
    case OperationCancelled = 2
    case OperationTimedOut = 3
    case OperationStartFailed = 4
    case InvalidIdentifiers = 5
    
    case BluetoothUnsupported = 100
    case BluetoothUnauthorized = 101
    case BluetoothPoweredOff = 102
    case BluetoothInUnknownState = 103
    case BluetoothResetting = 104
    case BluetoothStateChangeFailed = 105
    
    case DeviceConnectionFailed = 200
    case DeviceDisconnected = 201
    case DeviceRSSIReadFailed = 202
    case DeviceAlreadyConnected = 203
    case DeviceNotFound = 204
    case DeviceNotConnected = 205
    case DeviceMTUChangeFailed = 206
    
    case ServicesDiscoveryFailed = 300
    case IncludedServicesDiscoveryFailed = 301
    case ServiceNotFound = 302
    case ServicesNotDiscovered = 303
    
    case CharacteristicsDiscoveryFailed = 400
    case CharacteristicWriteFailed = 401
    case CharacteristicReadFailed = 402
    case CharacteristicNotifyChangeFailed = 403
    case CharacteristicNotFound = 404
    case CharacteristicsNotDiscovered = 405
    case CharacteristicInvalidDataFormat = 406
    
    case DescriptorsDiscoveryFailed = 500
    case DescriptorWriteFailed = 501
    case DescriptorReadFailed = 502
    case DescriptorNotFound = 503
    case DescriptorsNotDiscovered = 504
    case DescriptorInvalidDataFormat = 505
    
    case ScanStartFailed = 600
    case LocationServicesDisabled = 601
}


/// 蓝牙异常
struct RNBleError: Error {
    let errorCode: BleErrorCode
    let attErrorCode: Int?
    let iosErrorCode: Int?
    let reason: String?

    init(errorCode: BleErrorCode,
         reason: String? = nil,
         attErrorCode: Int? = nil,
         iosErrorCode: Int? = nil) {
        
        self.errorCode = errorCode
        self.reason = reason
        self.attErrorCode = attErrorCode
        self.iosErrorCode = iosErrorCode
    }
}

extension RNBleError {
    func asJSObject() -> [AnyHashable: Any] {
        return [
            "errorCode": self.errorCode.rawValue,
            "attErrorCode": self.attErrorCode.map {$0.description} ?? "null",
            "iosErrorCode": self.iosErrorCode.map {$0.description} ?? "null",
            "reason": self.reason ?? "null",
        ]
    }
}

File: android/src/main/java/com/jh/bluetooth/RNBleError.kt
package com.jh.bluetooth

import com.facebook.react.bridge.Arguments
import com.facebook.react.bridge.WritableMap

enum class BleErrorCode(val code: Int) {
    UnknownError(0),
    BluetoothManagerDestroyed(1),
    OperationCancelled(2),
    OperationTimedOut(3),
    OperationStartFailed(4),
    InvalidIdentifiers(5),

    BluetoothUnsupported(100),
    BluetoothUnauthorized(101),
    BluetoothPoweredOff(102),
    BluetoothInUnknownState(103),
    BluetoothResetting(104),
    BluetoothStateChangeFailed(105),

    DeviceConnectionFailed(200),
    DeviceDisconnected(201),
    DeviceRSSIReadFailed(202),
    DeviceAlreadyConnected(203),
    DeviceNotFound(204),
    DeviceNotConnected(205),
    DeviceMTUChangeFailed(206),

    ServicesDiscoveryFailed(300),
    IncludedServicesDiscoveryFailed(301),
    ServiceNotFound(302),
    ServicesNotDiscovered(303),

    CharacteristicsDiscoveryFailed(400),
    CharacteristicWriteFailed(401),
    CharacteristicReadFailed(402),
    CharacteristicNotifyChangeFailed(403),
    CharacteristicNotFound(404),
    CharacteristicsNotDiscovered(405),
    CharacteristicInvalidDataFormat(406),

    DescriptorsDiscoveryFailed(500),
    DescriptorWriteFailed(501),
    DescriptorReadFailed(502),
    DescriptorNotFound(503),
    DescriptorsNotDiscovered(504),
    DescriptorInvalidDataFormat(505),

    ScanStartFailed(600),
    LocationServicesDisabled(601),
}

/**
 * 错误信息类
 */
data class RNBleError(val errorCode: BleErrorCode,
                      val attErrorCode: Int?,
                      val iosErrorCode: Int?,
                      val reason: String?) : Error() {
    fun asJSObject(): WritableMap{
        val result = Arguments.createMap()
        result.putInt("errorCode",errorCode.code)
        result.putString("attErrorCode",(attErrorCode?:0).toString())
        result.putString("iosErrorCode",(iosErrorCode?:0).toString())
        result.putString("reason",(reason?:""))
        return result
    }
}

